### PR TITLE
Add support for setting column type in Clickhouse.

### PR DESCRIPTION
**Pull request by subkanthi**
Added support for setting column type in clickhouse
Relates to #15515

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# ClickHouse
* Add support for changing column types. ({issue}`15515`)
```

### CODE REVIEW

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 1]
```diff
@@ -459,7 +459,12 @@ public void setColumnComment(ConnectorSession session, JdbcTableHandle handle, J
     @Override
     public void setColumnType(ConnectorSession session, JdbcTableHandle handle, JdbcColumnHandle column, Type type)
     {
-        throw new TrinoException(NOT_SUPPORTED, "This connector does not support setting column types");
+        String sql = format(
+                "ALTER TABLE %s MODIFY COLUMN %s %s",
+                quoted(handle.asPlainTable().getRemoteTableName()),
+                quoted(column.getColumnName()),
+                toWriteMapping(session, type).getDataType());
+        execute(session, sql);
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:467]: Please remove the below lines from `BaseClickHouseConnectorTest`
```java
            case SUPPORTS_SET_COLUMN_TYPE:
                return false;
```

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 2]
```diff
@@ -459,7 +459,12 @@ public void setColumnComment(ConnectorSession session, JdbcTableHandle handle, J
     @Override
     public void setColumnType(ConnectorSession session, JdbcTableHandle handle, JdbcColumnHandle column, Type type)
     {
-        throw new TrinoException(NOT_SUPPORTED, "This connector does not support setting column types");
+        String sql = format(
+                "ALTER TABLE %s MODIFY COLUMN %s %s",
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:463]: ClikcHouse preserves other column definitions (e.g. default_expr, codec) when we omit them? I would recommend adding the test to `BaseClickHouseConnectorTest`.

Comment (subkanthi) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:463]: All the tables(nation, region, customer, orders) in the tpch database, seem to be of Engine family `Log` and `MODIFY TABLE` is not supported in Engine family Log in ClickHouse. 
Is it OK if I create another table with Engine family MergeTree

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 3]
```diff
@@ -2625,23 +2624,25 @@ public void testRenameColumn()
...
-        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_set_column_type_", "AS SELECT CAST(123 AS integer) AS col")) {
-            assertUpdate("ALTER TABLE " + table.getName() + " ALTER COLUMN col SET DATA TYPE bigint");
-
-            assertEquals(getColumnType(table.getName(), "col"), "bigint");
-            assertThat(query("SELECT * FROM " + table.getName()))
-                    .skippingTypesCheck()
-                    .matches("VALUES bigint '123'");
-        }
+        String tableName = "test_set_column_type";
+        assertUpdate("CREATE TABLE " + tableName + " (a bigint, b double NOT NULL, c varchar(50)) WITH (order_by=ARRAY['b'], engine = 'MergeTree')");
```
Comment (ebyhr) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2631]: Please don't put ClickHouse specific code to `BaseConnectorTest`. You can add a test method to `BaseClickHouseConnectorTest`.

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 4]
```diff
@@ -2625,23 +2624,25 @@ public void testRenameColumn()
...
+
+        assertUpdate("ALTER TABLE " + tableName + " ALTER COLUMN a SET DATA TYPE varchar(50)");
+
+        assertEquals(getColumnType(tableName, "a"), "varchar");
+        assertThat((String) computeScalar("show create table " + tableName)).contains("CREATE TABLE clickhouse.tpch.test_set_column_type (\n" +
```
Comment (ebyhr) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2636]: nit: Uppercase `show create table`. I would recommend verifying the table definition on ClikcHouse instead of Trino because the connector doesn't support all column definitions.

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 5]
```diff
@@ -2625,23 +2624,25 @@ public void testRenameColumn()
...
-
-            assertEquals(getColumnType(table.getName(), "col"), "bigint");
-            assertThat(query("SELECT * FROM " + table.getName()))
-                    .skippingTypesCheck()
-                    .matches("VALUES bigint '123'");
-        }
+        String tableName = "test_set_column_type";
+        assertUpdate("CREATE TABLE " + tableName + " (a bigint, b double NOT NULL, c varchar(50)) WITH (order_by=ARRAY['b'], engine = 'MergeTree')");
+
+        assertUpdate("ALTER TABLE " + tableName + " ALTER COLUMN a SET DATA TYPE varchar(50)");
```
Comment (ebyhr) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2633]: Column `a` doesn't have any definition. The new doesn't ensure if `SET DATA TYPE` preserved `a` column definition correctly or not because it created the table with table level property. I would recommend creating a table on Clickhouse so that we can specify column definitions (default_expr, codec, TTL).

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java (hunk 3 above)
Comment (subkanthi) [reply testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2631]: Sorry my mistake, still trying to understand the code layout.

**Comment by ebyhr** [comment]
Please feel free to ping me in Slack https://trino.io/slack.html if you need a help.

**Comment by subkanthi** [comment]
> Please feel free to ping me in Slack https://trino.io/slack.html if you need a help.

Hi @ebyhr , the `testSetColumnTypes` tests are failing , I couldn't figure out the right syntax to do a 
```
AS SELECT CAST() Engine=MergeTree
```
in ClickHouse, not even sure if its supported in CH, the parser fails because of the brackets. it works without Engine as it defaults to Log, but Log does not support MODIFY COLUMN.
Thoughts?

**Comment by ebyhr** [comment]
> I couldn't figure out the right syntax to do a
AS SELECT CAST() Engine=MergeTree

Why the cast doesn't contain value and type?  

By the way, please set code style configuration in your IDE. 
https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#code-style

**Comment by subkanthi** [comment]
> > I couldn't figure out the right syntax to do a
> > AS SELECT CAST() Engine=MergeTree
> 
> Why the cast doesn't contain value and type?
> 
> By the way, please set code style configuration in your IDE. https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#code-style

Hi @ebyhr ,
Question is how do I change the engine to MergeTree here
```
        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_set_column_type_", "AS SELECT CAST(123 AS integer) AS col")) {

```
This translates to the following CH create table syntax.
```
│ CREATE TABLE tpch.test_set_column_type_z65ft0194s
(
    `col` Nullable(Int32)
)
ENGINE = Log │

```

**Comment by ebyhr** [comment]
> how do I change the engine to MergeTree here

Please try `"WITH (engine = '...') AS SELECT ...` https://trino.io/docs/current/sql/create-table-as.html

**Comment by subkanthi** [comment]
> > how do I change the engine to MergeTree here
> 
> Please try `"WITH (engine = '...') AS SELECT ...` https://trino.io/docs/current/sql/create-table-as.html

Thanks, that helped a bit, but someone I need another column that is not null, since I can't alter modify an order_by column in ClickHouse.

```
            table = new TestTable(getQueryRunner()::execute, "test_set_column_type_", "WITH (engine='mergetree',order_by=ARRAY['col2']) AS SELECT CAST(" + setup.sourceValueLiteral() + " AS " + setup.sourceColumnType() + ") AS col, CAST(123 as int) AS col2");
```

```
io.trino.testing.QueryFailedException: Code: 44. DB::Exception: Sorting key cannot contain nullable columns. (ILLEGAL_COLUMN) (version 21.11.10.1 (official build))
```

**Comment by ebyhr** [comment]
Please use CREATE TABLE statement instead.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 6]
```diff
@@ -817,4 +922,201 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+                .add(new SetColumnTypeSetup("tinyint", "TINYINT '127'", "smallint"))
+                .add(new SetColumnTypeSetup("smallint", "SMALLINT '32767'", "integer"))
+                .add(new SetColumnTypeSetup("integer", "2147483647", "bigint"))
+                .add(new SetColumnTypeSetup("bigint", "BIGINT '-2147483648'", "integer"))
+                .add(new SetColumnTypeSetup("real", "REAL '10.3'", "double"))
+                .add(new SetColumnTypeSetup("real", "REAL 'NaN'", "double"))
+                .add(new SetColumnTypeSetup("decimal(5,3)", "12.345", "decimal(10,3)")) // short decimal -> short decimal
+                .add(new SetColumnTypeSetup("decimal(28,3)", "12.345", "decimal(38,3)")) // long decimal -> long decimal
+                .add(new SetColumnTypeSetup("decimal(5,3)", "12.345", "decimal(38,3)")) // short decimal -> long decimal
+                .add(new SetColumnTypeSetup("decimal(5,3)", "12.340", "decimal(5,2)"))
```
Comment (subkanthi) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:1043]: time, timestamp, addrow were removed since it looks like the translation to clickhouse is not implemented.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 7]
```diff
@@ -83,7 +93,7 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
                 return false;
 
             case SUPPORTS_SET_COLUMN_TYPE:
-                return false;
+                return true;
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:96]: Remove L95 & L96. It's enabled by default.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 8]
```diff
@@ -215,38 +225,24 @@ protected String tableDefinitionForAddColumn()
...
-
-            assertUpdate("INSERT INTO " + tableName + " VALUES ('a', 'b')", 1);
-
-            // ClickHouse set an empty character as the default value
-            assertUpdate("ALTER TABLE " + tableName + " ADD COLUMN c_varchar varchar NOT NULL");
-            assertFalse(columnIsNullable(tableName, "c_varchar"));
-            assertQuery("SELECT c_varchar FROM " + tableName, "VALUES ''");
-        }
-    }
+//    @Override
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:228]: Please remove this comment.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 9]
```diff
@@ -325,7 +321,7 @@ protected TestTable createTableWithDefaultColumns()
                         "col_nullable Nullable(Int64)," +
                         "col_default Nullable(Int64) DEFAULT 43," +
                         "col_nonnull_default Int64 DEFAULT 42," +
-                        "col_required2 Int64) ENGINE=Log");
+                        "col_required2 Int64) ENGINE=MergeTree() order by col_required2");
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:324]: nit: Uppercase `order by`

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 10]
```diff
@@ -686,6 +682,115 @@ public void testCharTrailingSpace()
         throw new SkipException("Implement test for ClickHouse");
     }
 
+    @Override
+    public void testNativeQuerySimple()
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:686]: Please revert unrelated changes.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 11]
```diff
@@ -817,4 +922,202 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
             return properties.buildOrThrow();
         }
     }
+
+    @Test
+    @Override
+    public void testSetColumnType()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_CREATE_TABLE_WITH_DATA));
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:930]: Remove this redundant line. The connector already knows their behavior.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 12]
```diff
@@ -817,4 +922,202 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
             return properties.buildOrThrow();
         }
     }
+
+    @Test
+    @Override
+    public void testSetColumnType()
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:928]: Please leave code comments why these test are overridden.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 13]
```diff
@@ -817,4 +922,202 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_CREATE_TABLE_WITH_DATA));
+        String tableName = "test_set_column_type";
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:931]: Add random suffix.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 14]
```diff
@@ -817,4 +922,202 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+    @Test
+    @Override
+    public void testSetColumnIncompatibleType()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_SET_COLUMN_TYPE) && hasBehavior(SUPPORTS_CREATE_TABLE_WITH_DATA));
+
+        String tableName = "test_set_column_incompatible_type";
+
+        assertUpdate("CREATE TABLE " + tableName + " (col bigint, col2 int not null) WITH (order_by=ARRAY['col2'], engine = 'MergeTree')");
+        assertUpdate("ALTER TABLE " + tableName + " ALTER COLUMN col SET DATA TYPE integer");
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:956]: The base test changes varchar type to integer. This overridden test is different from the base test purpose.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 15]
```diff
@@ -817,4 +922,202 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_SET_COLUMN_TYPE) && hasBehavior(SUPPORTS_CREATE_TABLE_WITH_DATA));
+
+        String tableName = "test_set_column_out_of_range";
+        assertUpdate("CREATE TABLE " + tableName + " (col bigint, col2 int not null) WITH (order_by=ARRAY['col2'], engine = 'MergeTree')");
+        query("insert into " + tableName + " values(9223372036854775807, 22)");
+        assertUpdate("ALTER TABLE " + tableName + " ALTER COLUMN col SET DATA TYPE integer");
+
+        assertEquals(getColumnType(tableName, "col"), "integer");
+        assertQuery("SELECT col FROM " + tableName, "VALUES -1");
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:972]: This looks like a bug. We may need to deny such operations. Could you find a GitHub issue of ClickHouse about this?

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 16]
```diff
@@ -817,4 +922,202 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+        }
+    }
+
+    @Test
+    @Override
+    public void testSetColumnTypeWithNotNull()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_SET_COLUMN_TYPE) && hasBehavior(SUPPORTS_NOT_NULL_CONSTRAINT));
+
+        String tableName = "test_set_column_with_not_null";
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:1008]: Add random suffix.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 17]
```diff
@@ -817,4 +922,202 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+        assertEquals(getColumnType(tableName, "col"), "integer");
+        assertQuery("SELECT col FROM " + tableName, "VALUES -1");
+    }
+
+    @Test
+    @Override
+    public void testSetColumnTypeWithComment()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_SET_COLUMN_TYPE) && hasBehavior(SUPPORTS_CREATE_TABLE_WITH_COLUMN_COMMENT));
+        String tableName = "test_set_column_with_comment";
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:980]: Add random suffix.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 18]
```diff
@@ -817,4 +922,202 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+        assertUpdate("CREATE TABLE " + tableName + " (col bigint not null, col2 int not null) WITH (order_by=ARRAY['col2'], engine = 'MergeTree')");
+        assertFalse(columnIsNullable(tableName, "col"));
+
+        assertUpdate("ALTER TABLE " + tableName + " ALTER COLUMN col SET DATA TYPE bigint");
+        assertFalse(columnIsNullable(tableName, "col"));
+    }
+
+    @DataProvider
+    @Override
+    public Object[][] setColumnTypesDataProvider()
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:1019]: Why do you need to override this method? Please override `filterSetColumnTypesDataProvider` instead.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 19]
```diff
@@ -817,4 +922,202 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+
+        assertEquals(getColumnType(tableName, "col"), setup.newColumnType());
+        assertThat(query("SELECT col FROM " + tableName))
+                .skippingTypesCheck()
+                .matches("SELECT " + setup.newValueLiteral());
+    }
+
+    @Test
+    @Override
+    public void testAddNotNullColumnToEmptyTable()
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:1078]: Revert changes for testAddNotNullColumnToEmptyTable & testAddNotNullColumn.

Comment (subkanthi) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:1078]: Hi @ebyhr , both the tests have to be overridden because it fails for the default Log Engine.

Comment (ebyhr) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:1078]: I knew. However, that is unrelated to the commit for changing column types.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java (hunk 18 above)
Comment (subkanthi) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:920]: overriding `filterSetColumnTypesDataProvider` doesnt seem to work.
Because the `setColumnTypeSetupData` is called from `setColumnTypesDataProvider`

```   @DataProvider
    public Object[][] setColumnTypesDataProvider()
    {
        return setColumnTypeSetupData().stream()
                .map(this::filterSetColumnTypesDataProvider)
                .flatMap(Optional::stream)
                .collect(toDataProvider());
    }
```

Comment (ebyhr) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:920]: Please take a look at other connectors. e.g. `TestIcebergParquetConnectorTest`

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java (hunk 15 above)
Comment (subkanthi) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:870]: `int` seems to map to `Int32` in ClickHouse, the max value in ClickHouse for `Int32` is `2147483647`.
```
┌─statement──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CREATE TABLE tpch.test_set_column_out_of_rangekgarmodrdz
(
    `col` Int32,
    `col2` Int32
)
ENGINE = MergeTree
ORDER BY col2
SETTINGS index_granularity = 8192 │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘


```

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 20]
```diff
@@ -817,4 +819,168 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
             return properties.buildOrThrow();
         }
     }
+
+    @Test
+    @Override
+    public void testSetColumnType()
+    {
+        String tableName = "test_set_column_type" + randomNameSuffix();
+        assertUpdate("CREATE TABLE " + tableName + " (a bigint, b double NOT NULL, c varchar(50)) WITH (order_by=ARRAY['b'], engine = 'MergeTree')");
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:828]: Please use `TestTable` instead. Other places are also the same.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 21]
```diff
@@ -817,4 +819,168 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+        assertUpdate("ALTER TABLE " + tableName + " ALTER COLUMN a SET DATA TYPE varchar(50)");
+        assertEquals(getColumnType(tableName, "a"), "varchar");
+        assertThat((String) computeScalar("show create table " + tableName)).contains("CREATE TABLE " + "clickhouse.tpch." + tableName + " (\n" +
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:831]: I don't understand why this assertion which doesn't exist in the base test method was added. Please extract a test method.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java (hunk 14 above)
Comment (ebyhr) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:852]: How ClickHouse handle values which can't cast to integer? e.g. `test string`
Probably, we shouldn't allow changing the type from varchar to numeric types in the connector.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 22]
```diff
@@ -817,4 +819,168 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+        assertUpdate("CREATE TABLE " + tableName + " (col bigint, col2 int not null) WITH (order_by=ARRAY['col2'], engine = 'MergeTree')");
+        assertUpdate("ALTER TABLE " + tableName + " ALTER COLUMN col SET DATA TYPE integer");
+        assertEquals(getColumnType(tableName, "col"), "integer");
+    }
+
+    @Test
+    @Override
+    public void testSetColumnOutOfRangeType()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_SET_COLUMN_TYPE) && hasBehavior(SUPPORTS_CREATE_TABLE_WITH_DATA));
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:860]: Remove redundant logic.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java (hunk 15 above)
Comment (ebyhr) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:867]: Let's deny changing such types.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 23]
```diff
@@ -817,4 +819,168 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+        assertUpdate("ALTER TABLE " + tableName + " ALTER COLUMN col SET DATA TYPE integer");
+        assertEquals(getColumnType(tableName, "col"), "integer");
+        assertQuery("SELECT col FROM " + tableName, "VALUES -1");
+    }
+
+    @Test
+    @Override
+    public void testSetColumnTypeWithComment()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_SET_COLUMN_TYPE) && hasBehavior(SUPPORTS_CREATE_TABLE_WITH_COLUMN_COMMENT));
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:874]: Remove.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 24]
```diff
@@ -817,4 +819,168 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+
+        assertUpdate("ALTER TABLE " + tableName + " ALTER COLUMN col SET DATA TYPE bigint");
+        assertEquals(getColumnComment(tableName, "col"), "test comment");
+    }
+
+    @Test
+    @Override
+    public void testSetColumnTypeWithDefaultColumn()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_SET_COLUMN_TYPE) && hasBehavior(SUPPORTS_INSERT));
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:888]: Remove.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 25]
```diff
@@ -817,4 +819,168 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+        assertFalse(columnIsNullable(tableName, "col"));
+
+        assertUpdate("ALTER TABLE " + tableName + " ALTER COLUMN col SET DATA TYPE bigint");
+        assertFalse(columnIsNullable(tableName, "col"));
+    }
+
+    @Override
+    protected Optional<SetColumnTypeSetup> filterSetColumnTypesDataProvider(SetColumnTypeSetup setup)
+    {
+        if (setup.sourceColumnType().equals("tinyint")) {
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:915]: Please use switch and leave the reason for those change to each block.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 26]
```diff
@@ -817,4 +819,168 @@ private Map<String, String> getTableProperties(String schemaName, String tableNa
...
+        }
+
+        return Optional.empty();
+    }
+
+    @Test(dataProvider = "setColumnTypesDataProvider")
+    @Override
+    public void testSetColumnTypes(SetColumnTypeSetup setup)
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_SET_COLUMN_TYPE) && hasBehavior(SUPPORTS_CREATE_TABLE_WITH_DATA));
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:968]: Remove.

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 27]
```diff
@@ -5864,7 +5864,7 @@ public void testMergeAllColumnsReversed()
         assertUpdate("DROP TABLE " + targetTable);
     }
 
-    private void verifyUnsupportedTypeException(Throwable exception, String trinoTypeName)
+    protected void verifyUnsupportedTypeException(Throwable exception, String trinoTypeName)
```
Comment (ebyhr) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:5867]: Revert unrelated change.

**Review by ebyhr — COMMENTED** [review]
Please squash commits into one.

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java (hunk 27 above)
Comment (subkanthi) [reply testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java]: done.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java (hunk 26 above)
Comment (subkanthi) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java]: Removed.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java (hunk 22 above)
Comment (subkanthi) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java]: Removed.

**Comment by mosabua** [comment]
:wave: @subkanthi @ebyhr  - this PR has become inactive. We hope you are still interested in working on it. Please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.
